service/storage_service.hh

```diff
@@ -911,6 +911,37 @@ private:
 public:
     future<bool> is_cleanup_allowed(sstring keyspace);
     bool is_repair_based_node_ops_enabled();
+    using node_operations_observer =  noncopyable_function<future<>(const node_ops_cmd_request& req, storage_service& ss)>;
+    void register_node_operation_observer(node_operations_observer observer);
+private:
+    std::vector<node_operations_observer> _node_operations_observers;
+
+    using wrapped_node_ops_cmd_response_processor = noncopyable_function<future<>(const gms::inet_address&, future<node_ops_cmd_response>)>;
+    using node_ops_cmd_response_processor = noncopyable_function<future<>(const gms::inet_address&, node_ops_cmd_response)>;
+
+    future<> send_node_ops_cmd(auto&& nodes,node_ops_cmd_request &req, node_ops_cmd_response_processor resp_process) {
```

Comment:
1. nitpick: missing space after `auto&& nodes,`

2. Unless I'm mistaken, `send_node_ops_cmd` and `send_node_ops_cmd_wrapped` are already only called from within a seastar thread. So instead of allocating data with `do_with`, it would be better to explicitly mark these functions with a "needs to be called from a seastar thread context" comment and write their code as if it was inside a thread (so there's no need to use `do_with`). Note that you already assume that one of the parameters is alive during the whole call (`node_ops_cmd_request &req`), and that's because it's implicitly known that these functions are only called from seastar threads anyway.

Reply:
Yes, it is a conceptual mistake on my part, I was trying to be prepared for the day this whole thread is co-routinized, 
however, it will be very similar in terms of objects life scope so I can do the change anyway :slightly_smiling_face: